src/Bicep.Core/TypeSystem/Providers/ProviderSpecificationSyntaxFactory.cs

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Text.RegularExpressions;
+using Bicep.Core.Parsing;
+
+namespace Bicep.Core.Syntax.Providers;
+
+public static partial class ProviderSpecificationSyntaxFactory
+{
+    // The setting below adds syntax highlighting for regex.
+    // language=regex
+    private const string NamePattern = "[a-zA-Z][a-zA-Z0-9]+";
+
+    private const string BicepRegistryAddressPattern = @"br[:\/]\S+";
+
+    // Regex copied from https://semver.org/.
+    // language=regex
+    private const string SemanticVersionPattern = @"(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?";
+
+    [GeneratedRegex(@$"^(?<name>{NamePattern})$", RegexOptions.ECMAScript | RegexOptions.Compiled)]
+    private static partial Regex ConfigManagedSpecificationPattern();
+
+    [GeneratedRegex(@$"^(?<name>{NamePattern})@(?<version>{SemanticVersionPattern})$", RegexOptions.ECMAScript | RegexOptions.Compiled)]
+    private static partial Regex LegacyBuiltInSpecificationPattern();
+
+    [GeneratedRegex(@$"^(?<address>{BicepRegistryAddressPattern})@(?<version>{SemanticVersionPattern})$", RegexOptions.ECMAScript | RegexOptions.Compiled)]
+    private static partial Regex InlinedSpecificationPattern();
+
+    [GeneratedRegex(@"^\S*[:\/](?<name>\S+)$", RegexOptions.ECMAScript | RegexOptions.Compiled)]
+    private static partial Regex RepositoryNamePattern();
+
+    public static IProviderSpecificationSyntax CreateProviderSpecificationSyntax(SyntaxBase syntax)
+     => syntax switch
+     {
+         StringSyntax stringSyntax when stringSyntax.TryGetLiteralValue() is { } value &&
+                                       TryCreateFromStringSyntax(stringSyntax, value) is { } specification
+             => specification,
+
+         IdentifierSyntax identifierSyntax
+             => new ConfigurationManagedProviderSpecificationSyntax(
+                 identifierSyntax.IdentifierName,
+                 IsValid: true,
+                 identifierSyntax.Span),
+
+         _ => new ProviderSpecificationSyntaxTrivia(syntax.Span)
```

Comment:
Where does `ProviderSpecificationSyntaxTrivia` get used? Could this instead be represented by an empty value (e.g. `null`)?

Reply:
the `Specification` property of the `ProviderDeclarationSyntax` class is expected to have a `IProviderSpecificationSyntax` object. This object is used when emitting a diagnostic and hence I am using the `ProviderSpecificationSyntaxTrivia` as a means to propagate the span information so the wiggly on the diagnostic is accurately represented.